### ISSUE TITLE
feature request: GlobalShortcuts portal

### ISSUE THREAD

**Issue by jokeyrhyme** [pop-os/xdg-desktop-portal-cosmic#4]
Howdie, thanks so much for sharing your work here

I noticed that there has been work done on specifying a portal for GlobalShortcuts: https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.GlobalShortcuts

This would be useful for communications apps that have push-to-talk functionality, as well as various other use cases

Do you know if this is likely to be on the roadmap? Or is it considered out-of-scope or otherwise not part of your project goals/vision?

Note that KDE does seem to advertise support for this portal: https://invent.kde.org/plasma/xdg-desktop-portal-kde/-/blob/master/data/kde.portal#L3

**Comment by Drakulix** [comment]
This might not be part of the first alpha, but we have been waiting for this portal to arrive. We plan to make this a central part of our key-bindings and key-customization story, so support is definitely on the roadmap rather sooner than later.

It is not yet clear, how the key bindings would be communicated back to the compositor (likely through some non-existent private wayland protocol), but if you want to contribute a implementation of the frontend to the portal, that would be very welcome.

**Comment by jokeyrhyme** [comment]
Yeah, the trick is allowing xdg-desktop-portal-cosmic to communicate with cosmic-comp in order to set the keyboard shortcut

hyprland did implement a custom wayland protocol to allow [xdg-desktop-portal-hyprland](https://github.com/hyprwm/xdg-desktop-portal-hyprland) to talk to [hyprland](https://github.com/hyprwm/Hyprland) for this purpose: https://github.com/hyprwm/hyprland-protocols/blob/main/protocols/hyprland-global-shortcuts-v1.xml

cosmic-comp could also expose a DBus API, or have a separate UNIX socket, 

Or (and I doubt this is the approach we'd take), xdg-desktop-portal-cosmic could directly modify cosmic-comp's config.ron, cosmic-comp could reload it after receiving an fs inotify or a SIGHUP or something

I did propose a general purpose solution that relies on DBus and shell scripts to make this work ( https://github.com/emersion/xdg-desktop-portal-wlr/issues/240#issuecomment-1546811352 ) but that wouldn't be necessary here seeing as xdg-desktop-portal-cosmic only has to support cosmic-comp (and vice versa)

**Comment by jokeyrhyme** [comment]
> It is not yet clear, how the key bindings would be communicated back to the compositor (likely through some non-existent private wayland protocol)

Could/should `cosmic-comp` subscribe to changes in `cosmic-config` the same way other components do, and have that be the way it learns when keyboard shortcuts have changed?

**Comment by jokeyrhyme** [comment]
Ah, I'm just realising that keybindings set via GlobalShortcuts are likely session-only, and probably shouldn't be persisted to disk, so `cosmic-config` would need a separate session-only concept if it were to be used for communicating GlobalShortcuts with other components

**Comment by Drakulix** [comment]
Right so there are multiple open questions on this subject.

Lets first define some names:
- "Default Shortcuts" - Shortcuts cosmic-comp sets by default and possible dependent of other configuration. E.g. how some shortcuts [already dependent on the workspace layout](https://github.com/pop-os/cosmic-comp/blob/master_jammy/src/config/key_bindings.rs#L214).
- "Static Shortcuts" - Shortcuts set by the user and expected to be persistent. Likely configured through cosmic-settings
- "Dynamic Shortcuts" - Shortcuts requested by an application through the global-shortcut Portal. These may or may not be persisted in some way (e.g. by automatically allowing applications to bind the exactly same shortcut again next time.)

Default Shortcuts are set by cosmic-comp and should likely include most of whats defined in the default-config file today. They could be overridden/replaced by static/dynamic shortcuts at runtime.

One set of open questions regards the "Static Shortcuts". They *should* be managed via cosmic-config for easy persistence and access for all components.

Open questions:
  - Should they be stored by cosmic-comp (and simply edited by cosmic-settings). How would something like binding media-keys to `MPRIS` work then? Would cosmic-comp start a cli tool and call a dbus-service or something like that?
  - Or should they be managed by cosmic-settings-daemon and on launch registered the same way as dynamic shortcuts through the portal are?
    - Whatever protocol is used here, this means the daemon would naturally get callbacks for all shortcuts and could directly issue dbus-requests and handle default apps / multiple implementations properly.
    - I am personally in favor of this approach, but that makes cosmic-comps shortcut customization story pretty bland on it's own and the portal or the settings-daemon practically a requirement in other environments. So this approach likely works best, if we use a commonly used protocol here, to make that dependency less jarring and easily replaceable.

The other set of questions concerns the dynamic shortcuts:
- How should communication work?
  - I don't see a point in adopting the hyprland protocol, as it is very compositor specific and not a lot of tooling exists around it. It might serve as a base for a cosmic-specific protocol though
  - The alternative would be the [action binder](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/216) protocol currently drafted upstream. The problem here is, that it has been long in the works, but isn't ready yet. So if we want to ship something based upon that now, we need to rename it under our own namespace, like we did for cosmic-screencopy, until it is merged upstream.

**Comment by Drakulix** [comment]
> Portal talks to compositor over D-Bus (`com.system76.CosmicComp.GlobalShortcuts`), following the `a11y_keyboard_monitor` pattern. Shortcut state is behind `Arc<parking_lot::RwLock>`: reads in the input hot path, writes from the D-Bus thread.

Note this isn't a pattern, it is simply how atspi (unfortunately) communicates. Our approach was based on a wayland-protocol + libei.

> **Questions:**
> 
>     1. Would you be open to accepting this as a contribution to Cosmic DE?
> 
>     2. If so, would you accept the D-Bus-based approach as a PR (two PRs, one per repo), or would you want it redone over a Wayland protocol first?

Unclear. This really is something that depends on our engineering team coming together and establishing a design on how this is supposed to work first, before any actual work can begin.

Imo my preferred approach would be to completely yank shortcut handling out of the compositor all together and using some upstream approach (like the unfortunately stalled action-binder protocol) for a separate shortcut-daemon to register input events it is interested in.

This would fit with the modularity of Cosmic in general and allow static (config files) and dynamic (portal) shortcuts to be established in a single component. Additionally a bunch of cosmic-components could stop having hardcoded default entries in cosmic-comps shortcut config, but register their shortcuts themselves. (Making them disappear automatically, if the user swaps out some component.)

But that is *just* my voice on the matter. If we would go with this route and additional code (e.g. to support a currently barely used GlobalShortcut portal) will make this more difficult.

So chances are nobody will have the resources to review your PR. Especially since it (necessarily) touches a bunch of components across our stack.

> PS: the implementation was AI-assisted (Claude Opus) with multiple review cycles covering design, security, language correctness, and architecture. All code is mine to contribute.

Note that we have pretty strict requirements on (any sort of) AI contributions. So this *might* push back review further depending of the quality of the code in question.

In short: I cannot tell you, as this is blocked on some design work. You probably could've spent your time better by reaching out first.

This isn't meant to insult your work and contributions in general are very welcome, but this sort of fundamental work is very difficult to externalize effectively. Especially since long-term maintenance is very important for our comparatively small team.

**Comment by kiroushi** [comment]
Thanks for the detailed response @Drakulix, and no offense taken at all.

To clarify the context: I built this purely as a personal workaround so I could use tools for Path of Exile like Sidekick and Awakened PoE Trade that need global shortcuts while gaming on Linux. The D-Bus approach was a pragmatic shortcut to get it working on my machine, not a design proposal for how Cosmic should handle this long-term.

No time was lost on my end since I'm using this daily regardless. I just wanted to check if it could be useful upstream and help other users in the same situation.

The separate shortcut-daemon idea with the action-binder protocol makes a lot of sense architecturally. Looking forward to seeing how that takes shape.

**Comment by lancemate** [comment]
> [@snoweuph](https://github.com/snoweuph) from my side I will digest Drakulix feedback into something more aligned with their vision and create a repo to sharing it with people like you. I don't expect this to materialize back into Cosmic, but at least linux gamers like us will be able to circumvent the issue while the definitive implementation comes.

Hi @kiroushi 
Just made the jump from gnome to cosmic but I’m lost without awakened poe trade. Is there any chance you can share your solution? Thanks bud.

**Comment by jdoss** [comment]
Not to distract from this dicussion, but I just opened two PRs which will fix Overlay programs like Awakened PoE Trade on COSMIC:

Smithay/smithay#1987
pop-os/cosmic-comp#2257

You can build cosmic-comp of my branch with [these](https://github.com/pop-os/cosmic-comp/pull/2257#issuecomment-4189751525) steps, and they can serve as a quick fix to unblock my fellow Exiles from switching to COSMIC.

**Comment by kiroushi** [comment]
@lancemate @jdoss thanks for the interest. Quick heads up: I recently became a parent so my time is pretty limited at the moment, but I'll try to find a spot in the coming weeks to clean things up and put something together for you folks.

@jdoss huge thanks for the parallel work. That's the use case that pushed me into this rabbit hole, so it's great to see fixes coming from that angle too. Fellow Exiles unite.

**Comment by jdoss** [comment]
@kiroushi No problem! As a not so new parent I want to say congrats and stay sane, Exile. 😄

**Comment by spartanatreyu** [comment]
Meta: Someone has made a GlobalShortcuts support tracking page for DEs / Applications.

https://areweglobalshortcutsyet.github.io/

**Comment by BluesAdam** [comment]
Oh man, to see "it will arrive sooner than later" back in 2023 to where we are today. I know this is yet another pain point caused by Wayland rather than Cosmic, but lacking such important and basic functionality really sucks.

I recently switched from KDE Plasma to Cosmic and this is the one thing I am truly missing. Is there any progress made in this or is it going to be worked on soon? Checked the roadmap page and didn't see anything relating to globalShortcuts portal.

I know it's easy to ask and hard to do, but I would really love to hear some good news on this front.

**Comment by KeronCyst** [comment]
This issue is butchering [CopyQ](https://hluk.github.io/CopyQ/)'s keyboard shortcut control:

Environment:
- COSMIC 1.2.0
- NixOS 26.05
- Wayland session
- CopyQ 13.0.0
- XDG_CURRENT_DESKTOP=COSMIC

Steps to reproduce:
1. Launch CopyQ with the tray/menu enabled.
2. Configure a global shortcut for “Show/hide main window” as Meta+V.
3. Keep the target application focused and press Meta+V.
4. Inspect CopyQ’s logs.

Expected:
The global shortcut should be registered through the Wayland portal and toggle the CopyQ window.

Actual:
CopyQ reports:
```
No such interface “org.freedesktop.portal.GlobalShortcuts” on object at path /org/freedesktop/portal/desktop
```
It also reports:
```
Failed to set global shortcut "Meta+V" for command "Show/hide main window".
```
The shortcut does not work. Invoking the same action manually through CopyQ’s D-Bus interface, for example with copyq show or copyq toggle, works normally.

This appears to be a missing GlobalShortcuts interface in xdg-desktop-portal-cosmic rather than a problem with COSMIC’s native custom shortcuts. Adding a native COSMIC shortcut is a useful workaround, but it does not help applications that need the standard Wayland portal interface.

For context, CopyQ was installed as a Nix package in:
/nix/store/18kc9kja0m4ili0d58ci4zg0n0645m29-CopyQ-13.0.0

The process is running in the COSMIC Wayland session, and CopyQ's log contains:
```
Warning: [copyq.globalshortcut] Failed to create portal global shortcuts session:
"No such interface “org.freedesktop.portal.GlobalShortcuts” on object at path /org/freedesktop/portal/desktop"
```

**Comment by KeronCyst** [comment]
Another reproducible issue on COSMIC 1.2.0 / NixOS 26.05 is KeePassXC's built-in auto-type feature. KeePassXC 2.7.12 is launched under X11/Xwayland with:
```
GDK_BACKEND=x11 QT_QPA_PLATFORM=xcb
```

The database is open and unlocked, an entry has Auto-Type enabled, and the global Auto-Type shortcut is configured. With another application focused, pressing the shortcut never opens KeePassXC's Auto-Type window and never types the credentials.

KeePassXC is using its X11/XCB Auto-Type plugin, but the global shortcut is not delivered to the X11 application. The KeePassXC failure reproduces on COSMIC regardless of whether the Auto-Type Settings page is visible. I initially suspected the missing page, but the actual problem is that Auto-Type has never functioned in this COSMIC/Xwayland environment.

**Comment by LiamillionSS** [comment]
Does keepass autotype even use globalshortcuts? That sounds like a completely separate issue.

As for "butchering" the keyboard shortcut control, thats because the portal implementation doesnt exist yet hence the issue still being open.

**Comment by andrew-verde** [comment]
I'm interested in contributing to this, following work on GameSentenceMiner's
Wayland shortcuts. Its portal client currently needs a manual compositor-command
fallback on COSMIC.

I've started a small standalone test client. It requests one shortcut and logs
press/release events while a separate browser window has focus. On Fedora 44,
with cosmic-comp 1.6.0 and xdg-desktop-portal-cosmic 1.6.0, the running portal
does not expose GlobalShortcuts. Successful delivery on a supporting desktop
still needs validation.

I read the March 17 feedback and would like to agree on scope before writing
compositor integration:

- Has the team settled on a separate daemon for static and dynamic shortcuts?
- If so, is action-binder the intended communication mechanism, including a
  temporary COSMIC protocol if needed?
- Would a standalone test client or a bounded part of the portal backend be
  useful now, and which part could the team realistically review?

The test client was developed with AI assistance using Codex. Any proposed
implementation would include that disclosure and a clear account of what was
tested.
